### ISSUE TITLE
Parse entire object for error array

### ISSUE THREAD

**Issue by boy51** [davidmdm/myzod#29]
Hi, to do form validations I need to be able to parse an entire object for errors (and not throw after the first error).

I would love to contribute myself but the way parse works for object type I don't see a way to do this without messing with the return type of parse, seriously changing ValidationError or writing a very hacky seperate method.

I think at least having an option like `abortEarly` for ObjectType.parse so that I have all error fields in `path` would make myzod suitable for form validation. You would just need to handle error messages yourself.

Is this something that is doable in the near future? If you have an implementation in mind I would love to contribute. Otherwise I would have to move to yup for my project which I wouldn't prefer doing.

**Comment by davidmdm** [comment]
I think it is doable. Also this hasn't been the first time I get this type of feature request for front-end form validation.
We just need to get the spec down.

My instinct would be to have a new option on the objectSchema.

```typescript
const personSchema = mz.object({ name: mz.string(), age: mz.number() }).collectErrors()
```

Then for the validation error we could add a new property where an object with keys(paths to error) to values validationErrors could be kept. The message for the root error could be a concatenation of the individual errors joined by newlines.

The issue is that objectType.parse is perhaps the most complicated and multi-pathed parse function to support records/key-signatures. So it might take some serious refactoring. And we don't want a serious penalty cost  in performance for the not collect errors case. 

If you want to take a stab at it, I would be super glad to work with you on this. Otherwise I can try and get on it this week.
Thanks for supporting me and not moving to yup immediately!

**Comment by boy51** [comment]
> The issue is that objectType.parse is perhaps the most complicated and multi-pathed parse function to support records/key-signatures. So it might take some serious refactoring. And we don't want a serious penalty cost in performance for the not collect errors case.

Definitely, and I can't say I understand most of what is happening there.

One of my first thoughts: since the object is always parsed in order of keys, ObjectType.parse could have an option like `startIndex` to make it start parsing from some index. That way there could be a method implementing a parse where each result is pushed into an array, and restarted from the index where it stopped, until it reaches the last key. 

Does that sound like it could work?

**Comment by davidmdm** [comment]
@boy51 I opened a PR. It is still very much incomplete as I have only added the collectErrors path for pure records. But maybe from this you might have enough of an idea of the direction I am going in? If you want to try your hand at contributing. Otherwise I'll try and be finished by end of week.

**Comment by boy51** [comment]
I'll check it out when I have some time :)

**Comment by davidmdm** [comment]
@boy51 
The implentation for object is done. I just need more tests and to write documentation. I was wondering if you would be down to help write some of those?

I was wondering if you needed the same `collectErrors` feature for arrays? Regardless if you do I might push that to a different PR. 

It's coming in soon!

also I managed to get about 5-10% increase in general object parsing performance. Not bad

**Comment by davidmdm** [comment]
Ok! I merged. Let me know if it works for you! I updated the docs. I'll keep the issue open for a a day or two as I wait for your feedback. Otherwise I'll close it. The feature was released as myzod v1.3.0

**Comment by boy51** [comment]
Awesome, I will definitely check it later today! Is there also a solution for error messages (having an error message object with all errors of parsed object that can be outputted to frontend)? Otherwise I would like to help improving those.

I like how myzod has a tiny package size, 21k vs yup 60k. I believe that must be because yup has errors included in the main package. They have plug-in locales of course. Do you think myzod could have plug-in errors and plug-in locales in the future? How do you want to do it?

> I was wondering if you needed the same collectErrors feature for arrays? Regardless if you do I might push that to a different PR.

Yes, but I'm not so clear how that would work - the error for the array schema would be a different one than the containing item, right? What if an array is part of my object?

**Comment by davidmdm** [comment]
The standard myzod.ValidationError now has a property `collectedErrors` which is an object with the key to the underlying validationError for that key. What exactly would you like to improve? Custom error messages? If that's it we can open another issue for that :) 

Also don't forget myzod is orders of magnitude faster than yup. Although on the front-end that doesn't matter too much.

**Comment by boy51** [comment]
I checked it out and it works great!

I'm aiming to use it together with formik and the implementation is pretty nice. Heres a [sandbox link](https://codesandbox.io/s/goofy-hopper-fd9vq). Maybe it would be useful to include some form validation example in docs?

So my concern with errors is, they look like this: `"error parsing object at path: \"number\" - expected number to be greater than or equal to 4 but got 1"`. They are not very user-friendly to output in a form. We have custom messages, but I have to `.split("-")` the `error.message` if I want to show it at all. And then how would I make a custom message for a type error (`.typeError("...")` in yup)?

Then long term there is the question of locales.

I don't know if your vision of this library includes all that, of course everyone is free to implement their own error message. But I think the scope of this package should include form validation. Would love to participate in new issue for this :)

Thanks for the fast response! I will recommend this library to everyone I know :D

**Comment by davidmdm** [comment]
@boy51 

Yeah, I am a backend developer by trade, so my use case was generally for an api-esque message. Or something easily interpreted for developers. 

There are some ideas I have. Like splitting into the message and the path. Having the raw reason, and then concatening the path if we need it, instead of having it be part of the message. 

Predicates allow for custom messages, and even you can pass a function which can return the error message, so in that sense, you can already support a local via a close for custom predicate failures. 

I suppose we can do the same for type errors. I haven't really used yup much. So I am all ears for suggestions that would make this library more appealing to front end developers.

I look forward to more of your issues!
Thanks for spreading the word. 

I might do a post on reddit to advertise this lib a little soon.

**Comment by davidmdm** [comment]
@boy51 I tried to add you as a collaborator so I could tag you in the new issue i made about user defined messages.
I kept the issue open but I published a new version of myzod v1.4.1 where you can define user errors for type errors.

```
schema.onTypeError(string | () => string)
```

I was hoping you could give me some feedback before I started writing some docs.

**Comment by boy51** [comment]
Hey, amazing work, I will check it later today and participate in the issue :) I wish I could invest more time into this project, but I dont have a lot of free time at the moment. I did deploy myzod to production recently so I think I can get those downloads up some more soon ;)
